contrib/99designs/gqlgen/tracer.go

```diff
@@ -0,0 +1,127 @@
+package gqlgen
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/99designs/gqlgen/handler"
+)
+
+type trace struct {
+	analyticsRate float64
+}
+
+func Tracer() handler.Option {
+	t := New()
+	traceOpt := handler.Tracer(t)
+	resolverOpt := handler.ResolverMiddleware(func(ctx context.Context, next graphql.Resolver) (res interface{}, err error) {
+		// Option 1 - Trace doesn't show up.
+		span, ctx := tracer.StartSpanFromContext(ctx, "Resolver")
+
+//		// Option 2 - Trace doesn't show up.
+//		parent, _ := tracer.SpanFromContext(ctx)
+//		span := tracer.StartSpan("Resolver", tracer.ChildOf(parent.Context()))
+//		ctx = tracer.ContextWithSpan(ctx, span)
+
+//		// Option 3 - Trace shows up, but child traces have incorrect parent.
+//		parent, _ := tracer.SpanFromContext(ctx)
+//		span := tracer.StartSpan("Resolver", tracer.ChildOf(parent.Context()))
+
+		defer span.Finish()
+		res, err = next(ctx)
+		return
+	})
```

Comment:
I've been investigating using the `ResolverMiddleware` to create spans around the user code. Unfortunately, I don't think it will pan out. It took me a while to figure out what was going wrong here. Eventually I discovered that the context we pass to `next` here is leaked out by gqlgen:

Code generated by gqlgen from generated.go, calling our middleware:
```
ctx = ec.Tracer.StartFieldResolverExecution(ctx, rctx)
resTmp, err := ec.ResolverMiddleware(ctx, func(rctx context.Context) (interface{}, error) {
	ctx = rctx // use context from middleware stack in children
	return ec.resolvers.Mutation().CreateTodo(rctx, args["input"].(NewTodo))
})
...
ctx = ec.Tracer.StartFieldChildExecution(ctx)
```
When we call `next(ctx)` with our context, the closure above is called, and the line `ctx = rctx` leaks our context out to be used later, after we've already `Finish()`ed the span. This seems to cause the whole trace to be bad, and it does not get sent. If I comment the line `ctx = rctx`, the original context is used for subsequent stuff, and the trace is sent correctly. We get, as I hoped, a trace with Resolver spans that closely resemble user code:
<img width="1007" alt="Screen Shot 2019-10-17 at 5 00 44 PM" src="https://user-images.githubusercontent.com/1819836/67052398-b11a5680-f103-11e9-979a-872ae3b65608.png">
This requires a change to gqlgen, although I suspect they did the weird context thing on purpose, so I'll continue looking at other methods.

Reply:
Let's keep it simple in the beginning and do what we can easily do. There is no point to struggle to add tracing for things that are outside of the already provided interface, especially for an alpha library.

For the first version, let's just do as we discussed. Have a root span which is the operation execution and its children, something like:
```
[ --------------------------------------- Operation ---------------------------------------]
  [--- Field Execution ---][--- Resolver Execution ---][--- Field Child Execution ---]
```
After we have that, we'll think about more to add later, potentially working with the authors. But we should have a very first basic version merged, that works like this. Even if the resolver execution includes _some_ (arguably neglectable) library code, it's little and covers the user code too.

While we do this, we may gather some notes and bullet points to bring up with them for potential improvements and come back to this works later based on how the upstream changes get applied.